File: 40-combination-sum-ii/combination-sum-ii.py
class Solution(object):
    def combinationSum2(self, candidates, target):
        candidates.sort()
        res = []

        def search(i, curr, total):
            if total==target:
                res.append(curr[:])
                return
            if total > target or i==len(candidates):
                return
            curr.append(candidates[i])
            search(i+1,curr, total + candidates[i]) 
            curr.pop()      
            while i+1< len(candidates) and candidates[i] == candidates[i+1]:
                i+=1
            search(i+1, curr, total)
        search(0,[],0)
        return res

File: 40-combination-sum-ii/combination-sum-ii.ts
function combinationSum2(candidates: number[], target: number): number[][] {
    candidates.sort((a, b) => a - b);
    let answer: number[][] = [];

    function backtracker(arr: number[], start: number, currentSum: number) {
        if (currentSum === target) {
            answer.push([...arr]);
            return;
        }

        for (let i = start; i < candidates.length; i++) {
            let sum = currentSum + candidates[i];
            if ((i === start || candidates[i - 1] != candidates[i]) && sum <= target) {
                arr.push(candidates[i]);
                backtracker([...arr], i + 1, sum);
                arr.pop();
            }
        }
    }

    backtracker([], 0, 0);
    return answer;
};